text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,682 @@
+be cryptographically secure due to `fork` sharing memory.
+
+This kind of functionality is sometimes called "random device", as a
+possible alternative name.
+
+#### PRNGs
+
+Only two user-space pseudo-random number generators are provided:
+[ChaCha] and [Xorshift][xorshift].
+
+This specifically means the ISAAC and ISAAC64 implementations will be
+moved out of `std::rand`. Compared to ISAAC*, the ChaCha RNG is much
+simpler, has had deeper cryptanalysis and is designed with SIMD
+optimisations in mind so is more suited to being part of a standard
+library.
+
+These two types can be deterministically seeded and would expose this
+functionality as inherent methods (not trait implementations), and
+would implement `Random` to "optimally" randomly seed themselves from
+another source of randomness.
+
+For example, the stable public interface for `ChaChaRng` (and
+`XorShiftRng`) will look like:
+
+```rust
+pub struct ChaChaRng { ... }
+
+impl ChaChaRng {
+    pub fn from_seed(seed: &[u32]) -> ChaChaRng { ... } // seed: [u32; 4] for XorshiftRng
+}
+
+impl Random<FullRange> for ChaChaRng {
+    ...
+}
+impl Reader for ChaChaRng {
+    ...
+}
+impl Rng for ChaChaRng {
+    ...
+}
+```
+
+These names are chosen to be somewhat vague: `ChaChaRng` is actually,
+currently, an implementation of the ChaCha20 algorithm, but the "20"
+value could be able to be customised backwards-compatibly in future,
+with default type parameters. (Similarly Xorshift is strictly
+Xorshift128, but has parameters that could be tweaked.) E.g. in
+hypothetical syntax
+
+```rust
+struct ChaChaRng<static N: u32 = 20> {
+    /* state */
+}
+let default: ChaChaRng = ...;
+let faster: ChaChaRng<8> = ...;
+```
+
+
+#### Alternatives
+
+As per the Alternatives section of the [free functions] discussion,
+these could be named `SecureRng`, `BalancedRng` and `WeakRng` (or have
+wrapper types of those names). However, the precedent is to not do
+this, both in other parts of the Rust library (e.g. `collections`) and
+in other random libraries, e.g. [C++11's `random`][cpp_random] and
+[D's `std.random`][d_random] both use the algorithms' names.
+
+Note, `XorShiftRng` is renamed to `XorshiftRng` since the "Xorshift"
+capitalisation is the correct one.
+
+### Thread RNG
+
+There will be one other RNG-implementer: the thread RNG. This RNG is
+stored in thread-local storage (TLS) to avoid having to reinitialise
+it each time, it will be defined something like
+
+```
+struct ThreadRngInner {
+    inner: ChaChaRng,
+    generated_entropy: usize,
+}
+pub struct ThreadRng {
+    inner: Rc<RefCell<ThreadRngInner>>
+}
+```
+
+so that the internal state is shared between all instances of the
+random number generator.
+
+The `generated_entropy` field will record the number of bytes output
+by the RNG, and the whole `ThreadRng` will be reseeded from the
+operating system once this passes a certain threshold.
+
+#### Alternatives
+
+Do not regularly reseed from the operating system: only the OS RNG is
```

Comment:
This is an alternative because it is just preserving the status quo of the current `std::rand`. It does sound like it's not necessary.